### ISSUE TITLE
async nested test harnesses (zalgo issue?)

### ISSUE THREAD

**Issue by andrewdeandrade** [tapjs/tapjs#102]
Currently, the `process` function will automatically call `this.end()` whenever there are no children left to process in a harness. This is fine for synchronous tests where all the child harnesses are defined before `this.process` is first called on the `process.nextTick`. 

However, if you have a circumstance where you want to create child harness asynchronously, then `this.process` will see that there are no children left and call `this.end` automatically before your next child harness can be registered. 

I'm happy to submit a pull request to fix this, but I wanted to see how others think this should be resolved. IMHO, the harness should only end if `this.end` is called explicitly by some outside function or when the plan count is reached. Merely adding a conditional here https://github.com/isaacs/node-tap/blob/a444c755a981997d8934648b3e165a738b07780d/lib/tap-harness.js#L90 in `Harness.prototype.process` that only calls `this.end` if `this._plan && this._testCount === this._plan` should be sufficient AFAICT, but I will confirm this.

**Comment by andrewdeandrade** [comment]
Here's an example that fails

``` javascript
var tap = require('tap');

tap.test('Harness async test support', function(t) {
  t.plan(3);

  t.ok(true, 'A');

  t.test('B', function(tt) {
    tt.plan(2);

    setTimeout(function(){
      tt.test('C', function(ttt) {
        ttt.plan(1);
        ttt.ok(true, 'D');
      });
    }, 50);

    setTimeout(function() {
      tt.test('E', function(ttt) {
        ttt.plan(1);
        ttt.ok(true, 'F');
      });
    }, 100);
  });

  setTimeout(function() {
    t.test('G', function(tt) {
      tt.plan(3);
      tt.test('H', function(ttt) {
        ttt.plan(1);
        ttt.ok(true, 'I');
      });
      tt.test('J', function(ttt) {
        ttt.plan(1);
        ttt.ok(true, 'K');
      });
      tt.test('L', function(ttt) {
        ttt.plan(1);
        ttt.ok(true, 'M');
      });
    });
  }, 200);
});
```

The output should be:

```
TAP version 13
# Harness async test support
ok 1 A
# B
# C
ok 2 D
# E
ok 3 F
# G
# H
ok 4 I
# J
ok 5 K
# L
ok 6 M
```

The following change to the end of the `this.process` method in `Harness` fixes this:

``` javascript
  //console.error("got current?", !!current)
  if (current) {
    current.on("end", this.process)
    current.emit("ready")
    //console.error("emitted ready")
    //console.error("_plan", this._plan, this.constructor.name)
  } else if (this._plan && this._plan === this._testCount) {
    //console.error("Harness process: no more left.  ending")
    if (this._endNice) {
      this._endNice()
    } else {
      this.end()
    }
  } else {
    this._started = false;
  }
```

However, I would prefer to change the `this._started` flag to something like `this._processing` throughout, since it obviously doesn't make sense semantically to have the `this._started` flag set to false at the end like in my above code change.

@isaacs your thoughts?

**Comment by isaacs** [comment]
Fixed by #103
